src/passes/stringify-walker.h

```diff
@@ -229,25 +235,82 @@ struct HashStringifyWalker : public StringifyWalker<HashStringifyWalker> {
 
   void addUniqueSymbol(SeparatorReason reason);
   void visitExpression(Expression* curr);
+  // Converts the idx from relative to the beginning of the program to
+  // relative to its enclosing function, and returns the name of its function
+  std::pair<uint32_t, Name> makeRelative(uint32_t idx) const;
+
+private:
+  // Contains the indices that mark the start of each function
+  std::set<uint32_t> funcIndices;
+  // Maps the start idx of each function to the function name
+  std::unordered_map<uint32_t, Name> idxToFuncName;
 };
 
+struct OutliningSequence {
+  unsigned startIdx;
+  unsigned endIdx;
+  Name func;
+
+  OutliningSequence(unsigned startIdx, unsigned endIdx, Name func)
+    : startIdx(startIdx), endIdx(endIdx), func(func) {}
+};
+
+using Substrings = std::vector<SuffixTree::RepeatedSubstring>;
+
 // Functions that filter vectors of SuffixTree::RepeatedSubstring
 struct StringifyProcessor {
-  static std::vector<SuffixTree::RepeatedSubstring>
-  dedupe(const std::vector<SuffixTree::RepeatedSubstring>& substrings);
+  static Substrings repeatSubstrings(std::vector<uint32_t>& hashString);
+  static Substrings dedupe(const Substrings& substrings);
   // Filter is the general purpose function backing subsequent filter functions.
   // It can be used directly, but generally prefer a wrapper function
   // to encapsulate your condition and make it available for tests
-  static std::vector<SuffixTree::RepeatedSubstring>
-  filter(const std::vector<SuffixTree::RepeatedSubstring>& substrings,
-         const std::vector<Expression*> exprs,
-         std::function<bool(const Expression*)> condition);
-  static std::vector<SuffixTree::RepeatedSubstring>
-  filterLocalSets(const std::vector<SuffixTree::RepeatedSubstring>& substrings,
-                  const std::vector<Expression*> exprs);
-  static std::vector<SuffixTree::RepeatedSubstring>
-  filterBranches(const std::vector<SuffixTree::RepeatedSubstring>& substrings,
-                 const std::vector<Expression*> exprs);
+  static Substrings filter(const Substrings& substrings,
+                           const std::vector<Expression*>& exprs,
+                           std::function<bool(const Expression*)> condition);
+  static Substrings filterLocalSets(const Substrings& substrings,
+                                    const std::vector<Expression*>& exprs);
+  static Substrings filterLocalGets(const Substrings& substrings,
+                                    const std::vector<Expression*>& exprs);
+  static Substrings filterBranches(const Substrings& substrings,
+                                   const std::vector<Expression*>& exprs);
+};
+
+struct ReconstructStringifyWalker
+  : public StringifyWalker<ReconstructStringifyWalker> {
+  enum ReconstructState {
```

Comment:
It would be good to have comments here explaining what these states mean.

Reply:
done